### PR TITLE
net: gracefully disconnect better from remote

### PR DESCRIPTION
**Pull request by connor4312**
This adds in a step to gracefully disconnect (send a "disconnect"
message and then await the flush) before closing the workbench. In some
cases, like Remote - SSH, sending messages is more async than others. In
the exec server connection we handle a `zlib.flate` stream to compress
data to and from the VS Code server, and its API is asynchronous, so
this lets us ensure the stream is drained before giving the go-ahead
to close up shop

This lifecycle phase is a little awkward and I ended it putting it
directly in the lifecycle service: we don't want to do this in
`onWillShutdown` because some contributions want to save data to the
remote workspace, and if shutdown is vetoed it would leave a broken
state. But `onDidShutdown` is synchronous and already depended upon by
several other points, and changing that also felt a bit risky.

cc @alexdima

Refs #211462, will require some small adoption in Remote - SSH to close.

### CODE REVIEW

File: src/vs/workbench/services/lifecycle/common/lifecycleService.ts [hunk 1]
```diff
@@ -3,13 +3,13 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-import { Emitter } from 'vs/base/common/event';
 import { Barrier } from 'vs/base/common/async';
```
Comment (bpasero) [inline src/vs/workbench/services/lifecycle/common/lifecycleService.ts:6]: Unrelated change

File: src/vs/workbench/services/lifecycle/common/lifecycleService.ts [hunk 2]
```diff
@@ -44,7 +44,7 @@ export abstract class AbstractLifecycleService extends Disposable implements ILi
 
 	constructor(
 		@ILogService protected readonly logService: ILogService,
-		@IStorageService protected readonly storageService: IStorageService
+		@IStorageService protected readonly storageService: IStorageService,
```
Comment (bpasero) [inline src/vs/workbench/services/lifecycle/common/lifecycleService.ts:47]: Unrelated change

File: src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts [hunk 3]
```diff
@@ -66,6 +69,10 @@ export class NativeLifecycleService extends AbstractLifecycleService {
 			// trigger onWillShutdown events and joining
 			await this.handleWillShutdown(reply.reason);
 
+			// now that all services have stored their data, it's safe to terminate
+			// the remote connection gracefully before synchronously saying we're shut down
+			await this.handleRemoteAgentDisconnect();
```
Comment (bpasero) [inline src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts:74]: This seems like a strange dependency to me, why can the remote agent service not prolong the shutdown joining the event?

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts (hunk 3 above)
Comment (connor4312) [reply src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts:74]: In the PR description I mentioned why the two existing points aren't good fits:

> we don't want to do this in
> onWillShutdown because some contributions want to save data to the
> remote workspace, and if shutdown is vetoed it would leave a broken
> state. But onDidShutdown is synchronous and already depended upon by
> several other points, and changing that also felt a bit risky [and would churn a lot]

Comment (connor4312) [reply src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts:74]: Oh, nevermind., rereading today I see that shutdown can only be vetoed on beforeShutdown, not willShutdown. I can adjust that. I think the risk is if anyone in `onWillShutdown` has to do things like saving state to the remote agent. For example, the edit sessions contributions saves editors and interacts with the filesystem in order to generate its working copies. Maybe we introduce some kind of ordering in the joiners?

Comment (bpasero) [reply src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts:74]: I would not be opposed to having some kind of priority in the `join` method that would allow contributions to signal whether they have dependencies on other things and thus want early handling vs. late handling. We could simply split up the promise waiters here based on that:

https://github.com/microsoft/vscode/blob/8cb093f3643daf81765d8200afaa602eb03a1bb4/src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts#L165-L171

Comment (connor4312) [reply src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts]: Added and cleaned up :)

**Comment by bpasero** [comment]
If you don't mind, I might want to make it a bit more abstract and less specific for remote.

**Comment by connor4312** [comment]
Sure, do you have any better naming in mind?

**Comment by bpasero** [comment]
I would not encode the term "remote" in it, rather some "high" / "low" priority terms.

File: src/vs/workbench/services/lifecycle/test/electron-sandbox/lifecycleService.test.ts [hunk 4]
```diff
@@ -155,5 +156,32 @@ suite('Lifecycleservice', function () {
...
+		disposables.add(lifecycleService.onWillShutdown(e => {
+			e.join(async () => {
+				order.push('disconnect start');
+				await timeout(1);
+				order.push('disconnect end');
+			}, { id: 'test', label: 'test', order: WillShutdownJoinerOrder.Last });
+
+			e.join((async () => {
+				order.push('default start');
+				await timeout(1);
```
Comment (bpasero) [inline src/vs/workbench/services/lifecycle/test/electron-sandbox/lifecycleService.test.ts:171]: If we are using `timeout` I think we should also use fake timers to reduce flakyness.

File: src/vs/workbench/services/lifecycle/common/lifecycle.ts [hunk 5]
```diff
@@ -65,9 +65,31 @@ export interface BeforeShutdownErrorEvent {
...
+	readonly order?: WillShutdownJoinerOrder;
+}
+
+export interface IWillShutdownEventDefaultJoiner extends IWillShutdownEventJoiner {
+	readonly order?: WillShutdownJoinerOrder.Default;
+}
+
+export interface IWillShutdownEventLastJoiner extends IWillShutdownEventJoiner {
+	readonly order: WillShutdownJoinerOrder.Last;
 }
```
Comment (bpasero) [inline src/vs/workbench/services/lifecycle/common/lifecycle.ts:94]: If `Last` order is used, we must require a `() => Promise`, otherwise we cannot guarantee last execution. As for the other case, we can support just `Promise` as before.

File: src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts [hunk 6]
```diff
@@ -155,19 +155,24 @@ export class NativeLifecycleService extends AbstractLifecycleService {
...
-				promise.finally(() => pendingJoiners.delete(joiner));
+
+				if (joiner.order === WillShutdownJoinerOrder.Last) {
+					const promiseFn = typeof promiseOrPromiseFn === 'function' ? promiseOrPromiseFn : () => promiseOrPromiseFn;
+					lastJoiners.push(() => promiseFn().finally(() => pendingJoiners.delete(joiner)));
+				} else {
+					const promise = typeof promiseOrPromiseFn === 'function' ? promiseOrPromiseFn() : promiseOrPromiseFn;
+					promise.finally(() => pendingJoiners.delete(joiner));
+					joiners.push(promise);
+				}
```
Comment (bpasero) [inline src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts:175]: Please let me know if this still makes sense to you, I felt it was easier to understand.

**Review by bpasero — CHANGES_REQUESTED** [review]
@connor4312 pushed some cleanup and left comments, please check it out.

File: src/vs/workbench/services/lifecycle/test/electron-sandbox/lifecycleService.test.ts (hunk 4 above)
Comment (connor4312) [reply src/vs/workbench/services/lifecycle/test/electron-sandbox/lifecycleService.test.ts:171]: This is more just to have some async to verify that the code waits for the promise to resolve in the first stage before starting the last stage. If this does flake that means there is a bug in the implementation: the order of these events should be deterministic.

**Comment by connor4312** [comment]
Lgtm

File: src/vs/workbench/services/lifecycle/common/lifecycle.ts [hunk 7]
```diff
@@ -65,9 +65,31 @@ export interface BeforeShutdownErrorEvent {
...
 
+export enum WillShutdownJoinerOrder {
+
+	/**
+	 * Joiners to run before the `Last` joiners.
+	 */
+	Default = 1,
+
+	/**
+	 * The joiners to run last.
```
Comment (connor4312) [inline src/vs/workbench/services/lifecycle/common/lifecycle.ts:76]: It might be useful to add a comment that the workbench will be disconnecting when this happens. A new consumer of joiners doesn't have any other guidance which is the right one to use for their case, and accidentally using a joiner that requires the workspace in `Last` would be easy to miss in testing.

File: src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts [hunk 8]
```diff
@@ -180,6 +185,7 @@ export class NativeLifecycleService extends AbstractLifecycleService {
 
 		try {
 			await raceCancellation(Promises.settled(joiners), cts.token);
+			await raceCancellation(Promises.settled(lastJoiners.map(lastJoiner => lastJoiner())), cts.token);
```
Comment (connor4312) [inline src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts:188]: This change means that an rejection thrown in the Default priority joiners will cause the last joiners to not run. This should be pretty rare but we should handle this case

**Review by bpasero — APPROVED** [review]

**Comment by connor4312** [comment]
lgtm!

**Review by TylerLeonhardt — APPROVED** [review]
